rust/cubestore/cubestore/src/cachestore/lazy.rs

```diff
@@ -0,0 +1,168 @@
+    ) -> Result<Arc<Self>, CubeError> {
+        let store = RocksCacheStore::load_from_dump(path, dump_path, metastore_fs, config).await?;
+
+        Ok(Arc::new(Self {
+            init_signal: None,
+            state: tokio::sync::RwLock::new(LazyRocksCacheStoreState::Initialized { store }),
+        }))
+    }
+
+    pub async fn load_from_remote(
+        path: &str,
+        metastore_fs: Arc<dyn MetaStoreFs>,
+        config: Arc<dyn ConfigObj>,
+    ) -> Result<Arc<Self>, CubeError> {
+        let (init_flag, init_signal) = tokio::sync::watch::channel::<bool>(false);
+
+        Ok(Arc::new(Self {
+            init_signal: Some(init_signal),
+            state: tokio::sync::RwLock::new(LazyRocksCacheStoreState::FromRemote {
+                path: path.to_string(),
+                metastore_fs,
+                config,
+                init_flag,
+            }),
+        }))
+    }
+
+    async fn init(&self) -> Result<Arc<RocksCacheStore>, CubeError> {
+        {
+            let guard = self.state.read().await;
+            if let LazyRocksCacheStoreState::Initialized { store } = &*guard {
+                return Ok(store.clone());
+            };
+        }
+
+        let mut guard = self.state.write().await;
+        match &*guard {
+            LazyRocksCacheStoreState::FromRemote {
+                path,
+                metastore_fs,
+                config,
+                // receiver will be closed on drop
+                init_flag: _,
+            } => {
+                let store =
+                    RocksCacheStore::load_from_remote(&path, metastore_fs.clone(), config.clone())
+                        .await?;
+
+                *guard = LazyRocksCacheStoreState::Initialized {
+                    store: store.clone(),
+                };
+
+                Ok(store)
+            }
+            _ => Err(CubeError::internal(
+                "Unable to initialize Cache Store on lazy call, unexpected state".to_string(),
+            )),
+        }
+    }
+
+    async fn get_initialized_store(&self) -> Result<Arc<RocksCacheStore>, CubeError> {
+        let guard = self.state.read().await;
+        if let LazyRocksCacheStoreState::Initialized { store } = &*guard {
+            Ok(store.clone())
+        } else {
+            Err(CubeError::internal(
+                "Unable to extract store from the state".to_string(),
+            ))
+        }
+    }
+
+    pub async fn wait_upload_loop(&self) {
+        if let Some(init_signal) = &self.init_signal {
+            let _ = init_signal.clone().changed().await;
+        }
+
+        trace!("wait_upload_loop unblocked, Cache Store was initialized");
+
+        self.get_initialized_store()
+            .await
+            .unwrap()
+            .wait_upload_loop()
+            .await;
+    }
+
+    pub async fn stop_processing_loops(&self) {
+        if let Ok(store) = self.get_initialized_store().await {
+            store.stop_processing_loops().await;
```

Comment:
Should I introduce a new state (`Closed`) to protect a problem when someone calls `CacheStore->init` after`stop_processing_loops`?